Fungible token example

ailisp opened the pull request.

itegulov commented on examples/fungible-token/src/index.js:
```diff
@@ -0,0 +1,78 @@
+        let senderId = near.predecessorAccountId()
+        this.internalTransfer(senderId, receiverId, amount, memo)
+        let onTransferRet = near.jsvmCall(receiverId, 'ftOnTransfer', [senderId, amount, msg, receiverId])
+        // In JS, do not need a callback, ftResolveTransfer after ftOnTransfer Returns
+        // If any logic after ftOnTransfer Returns is required, just do it on onTransferRet.
```
Hmm, I am confused about how is JS SDK here different from Rust SDK. Could you elaborate on why this is not needed?

ailisp replied:
In Rust, cross contract call is async, so you cannot write sequential code after the cross contract call returns. Instead, you need prepare a callback function `ft_on_transfer`, and pass that to cross contract call `promise_then`. 

In JS, the call is sync, so just write code after it. You don't have to organize it in a callback function. In the simplest case, that is don't do anything with the return value here, just leave and ignore the value

itegulov reviewed (commented):
Did a quick pass, but still need to run this myself

ailisp commented on sdk/collections/lookup-map.js:
```diff
@@ -0,0 +1,33 @@
+
+    get(key) {
+        let storageKey = this.keyPrefix + key
+        return near.jsvmStorageRead(storageKey)
```
Yes. The other way is to have a get and getRaw like rust. This is actually `getRaw`. But I do want to decouple deserialization with the actual data structure as much as possible, so that 1) user can choose their own serialization; 2) the implementation of each data structure is simply the data structure part.

I think a proper way to do this need more discuss and consideration on trade offs. One possible solution is to let user inherit LookupMap to set their serialization. In this approach we may also need to provide some default impl, a JSONLookupMap, for example. The other approach is to make serialize as mixin, this is closest to Rust's traits.

But no matter which more well-designed approach we'll use, the greatest common divisor version can't be wrong. So I decide on this version for our first release.

ailisp commented on sdk/collections/lookup-map.js:
```diff
@@ -0,0 +1,33 @@
+
+    set(key, value) {
+        let storageKey = this.keyPrefix + key
+        return near.jsvmStorageWrite(storageKey, value)
```
Good catch!

ailisp replied:
Fixed

ailisp commented on sdk/near-bindgen.js:
```diff
@@ -16,7 +16,7 @@ export function NearBindgen (Class) {
-        let ret = new OriginalClass()
+        let ret = Object.create(NewClass.prototype)
         return ret
     }
```
I'm also not expert on JS and current solution is learned from limited resources on JS decorators out there. Will try out your approach!

ailisp replied:
Doesn't work, got error "WebAssembly trap: An `unreachable` opcode was executed.". Same after change `let ret = Object.create(NewClass.prototype)` to `OriginalClass.prototype`.

This is kind of hard to debug, usually if something in JavaScript wrong, the error is `smart contract panicked` plus a traceback. And I can look trackback to identify error in JS file. But this error doesn't. Maybe some one with more knowledge in JS can tell what goes wrong. Your suggested snippet is certainly more elegant.

ailisp commented on examples/fungible-token/src/index.js:
```diff
@@ -0,0 +1,78 @@
+    if (b) {
+        return
+    } else {
+        throw Error("assertion failed: " + str)
```
Host panic works. We use `throw new Error` is because more idiomatic in JavaScript. It is then converted to host panic call in `jsvm.c` so the effect is same.

ailisp commented:
@itegulov I've fixed or addressed your comments. Probably they're still not perfect, but we would need some working collections and examples before deadline. I'll merge this for now. Feel free to post a PR to enhance (I'll do so too). If there's nothing looks very wrong, let's do after the first release
